### PR TITLE
Check for existing Pod Clusters on creation

### PR DESCRIPTION
**Pull request by anthonybishopric**
This adds a required Session argument to pod cluster creation, ensuring that the pod cluster for a given 3-ple of pod ID, availability zone, and cluster name does not already exist.

### CODE REVIEW

File: pkg/kp/pcstore/consul_store.go [hunk 1]
```diff
@@ -44,9 +45,24 @@ func (s *consulStore) Create(
 	clusterName fields.ClusterName,
 	podSelector klabels.Selector,
 	annotations fields.Annotations,
+	session Session,
 ) (fields.PodCluster, error) {
 	id := fields.ID(uuid.New())
 
+	unlocker, err := s.lockForUpdateCreation(podID, availabilityZone, clusterName, session)
```
Comment (anthonybishopric) [inline pkg/kp/pcstore/consul_store.go]: I just realized this was named this way after "rolling updates" not "update or creation"

File: pkg/kp/pcstore/consul_store.go [hunk 2]
```diff
@@ -143,6 +159,42 @@ func (s *consulStore) pcPath(pcID fields.ID) (string, error) {
 	return path.Join(podClusterTree, pcID.String()), nil
 }
 
+func (s *consulStore) lockForUpdateCreation(podID types.PodID,
```
Comment (mpuncel) [inline pkg/kp/pcstore/consul_store.go]: probably should just be `lockForCreation`

File: pkg/kp/pcstore/consul_store.go [hunk 3]
```diff
@@ -44,9 +45,24 @@ func (s *consulStore) Create(
...
+		return fields.PodCluster{}, err
+	}
+	defer unlocker.Unlock()
+
+	existing, err := s.FindWhereLabeled(podID, availabilityZone, clusterName)
+	if err != nil {
+		return fields.PodCluster{}, util.Errorf("Couldn't determine if pod cluster exists already: %v", err)
+	}
+	if len(existing) > 0 {
+		return existing[0], util.Errorf("Pod cluster already exists for %v", pcCreateLockPath(podID, availabilityZone, clusterName))
```
Comment (mpuncel) [inline pkg/kp/pcstore/consul_store.go]: maybe give the ID of the pod cluster too?

Comment (mpuncel) [inline pkg/kp/pcstore/consul_store.go]: i guess that forces you to deal with `len(existing)` which is annoying so feel free to ignore

File: pkg/kp/pcstore/consul_store.go [hunk 4]
```diff
@@ -143,6 +159,42 @@ func (s *consulStore) pcPath(pcID fields.ID) (string, error) {
...
+	availabilityZone fields.AvailabilityZone,
+	clusterName fields.ClusterName) string {
+	return path.Join(kp.LOCK_TREE, podID.String(), availabilityZone.String(), clusterName.String())
+}
+
+func (s *consulStore) FindWhereLabeled(podID types.PodID,
+	availabilityZone fields.AvailabilityZone,
+	clusterName fields.ClusterName) ([]fields.PodCluster, error) {
+
+	sel := klabels.Everything().
```
Comment (mpuncel) [inline pkg/kp/pcstore/consul_store.go]: we should probably put the label set in an accessible place and then use `labelSet.AsSelector()`

File: pkg/kp/pcstore/consul_store_test.go [hunk 5]
```diff
@@ -217,3 +236,22 @@ func consulStoreWithFakeKV() *consulStore {
 		applicator: labels.NewFakeApplicator(),
 	}
 }
+
+type fakeSession struct {
```
Comment (mpuncel) [inline pkg/kp/pcstore/consul_store_test.go]: any chance of reusing `fakeSession` from `pkg/kp/kptest`?

Comment (mpuncel) [inline pkg/kp/pcstore/consul_store_test.go]: i dont' feel strongly

File: pkg/kp/pcstore/consul_store.go (hunk 3 above)
Comment (anthonybishopric) [inline pkg/kp/pcstore/consul_store.go]: Easy to add the first one at least

**Comment by mpuncel** [comment]
looks great just nits 👍

File: pkg/kp/pcstore/consul_store_test.go (hunk 5 above)
Comment (anthonybishopric) [inline pkg/kp/pcstore/consul_store_test.go]: Didn't realize it existed, will remove this one

File: pkg/kp/pcstore/pcstoretest/fake_pcstore.go [hunk 6]
```diff
@@ -56,3 +57,19 @@ func (p *FakePCStore) Delete(id fields.ID) error {
 	delete(p.podClusters, id)
 	return nil
 }
+
+func (p *FakePCStore) FindWhereLabeled(
```
Comment (mpuncel) [inline pkg/kp/pcstore/pcstoretest/fake_pcstore.go]: might be worth putting a `labels.NewFakeApplicator()` onto the fake store and actually write/delete/query labels so if the label set ever diverges from the fields the behavior will stay correct, but not necessary today.

Comment (mpuncel) [inline pkg/kp/pcstore/pcstoretest/fake_pcstore.go]: or maybe this fake store could just be the real store but with a fake consul api client put in? that'd be cool but can be a different PR

Comment (anthonybishopric) [inline pkg/kp/pcstore/pcstoretest/fake_pcstore.go]: I like the idea of using a fake applicator, since there will be needs for a fake pcstore. Will save for a future change though.

**Comment by mpuncel** [comment]
👍
